lumigator/backend/backend/services/jobs.py

```diff
@@ -297,17 +297,32 @@ def get_upstream_job_status(self, job_id: UUID) -> str:
             raise JobUpstreamError("ray", "error getting Ray job status", e) from e
 
     def get_job_logs(self, job_id: UUID) -> JobLogsResponse:
-        db_logs = self.job_repo.get(job_id)
-        if not db_logs:
-            raise JobNotFoundError(job_id, "Failed to find the job record holding the logs") from None
-        elif not db_logs.logs:
-            ray_db_logs = self.retrieve_job_logs(job_id)
-            self._update_job_record(job_id, logs=ray_db_logs.logs)
-            return ray_db_logs
-        else:
-            return JobLogsResponse(logs=db_logs.logs)
+        """Retrieves the logs for a job from the upstream service.
+
+        :param job_id: The ID of the job to retrieve logs for.
+        :return: The logs for the job.
+        :raises JobNotFoundError: If the job cannot be found.
+        :raises JobUpstreamError: If there is an error with the upstream service returning the job logs,
+                and there are no logs currently persisted in Lumigator's storage.
+        """
+        job = self.job_repo.get(job_id)
+        if not job:
+            raise JobNotFoundError(job_id) from None
+
+        try:
+            job_logs = self._retrieve_job_logs(job_id)
+        except JobUpstreamError:
+            # If we have logs stored, just return them to support 'offline' Ray
+            if job.logs:
+                return JobLogsResponse(logs=job.logs)
+            raise
+
+        # Update the database with the latest logs
+        self._update_job_record(job_id, logs=job_logs.logs)
```

Comment:
nit: This will do a db request to update the record every time `get_job_logs` gets called which might not be necessary since the logs may have not changed, might be better to check if the logs are different from `job.logs` first to avoid unnecessary requests/load on the db

I don't have enough context, but I'm under the impression that this is a bit sketchy architecture wise because the method name is "get_job_logs" so I would expect it to read only instead of updating the db/job after getting it, not sure if there's a better way to separate the reads from the writes, since you don't necessarily want to write to the db every time someone wants to fetch the logs.